pkg/telemetryquerylanguage/tql/compare.go

```diff
@@ -0,0 +1,334 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tql // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/telemetryquerylanguage/tql"
+
+import (
+	"bytes"
+
+	"golang.org/x/exp/constraints"
+)
+
+// The functions in this file implement a general-purpose comparison of two
+// values of type any, which for the purposes of TQL mean values that are one of
+// int, float, string, bool, or pointers to those, or []byte, or nil.
+
+// invalidComparison returns false for everything except NE (where it returns true to indicate that the
+// objects were definitely not equivalent).
+// It also gives us an opportunity to log something first if we want to do that.
```

Comment:
I've done it, but I was worried about the cost of just throwing something in here. I think we should make ParseQueries into a member function on a TQLParser type that you need to construct before you call ParseQueries(). Then it would be easy to add a caller-controlled logger to that object and use it here.

Reply:
Yes I agree. The issue for tracking adding a logger is https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9730